### PR TITLE
Backend app/feature/oauth

### PR DESCRIPTION
**Pull request by kiselev-danil**

### CODE REVIEW

File: effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/TokenVerifier.kt [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            if ((acceptableMailDomain == hostedDomain) && emailVerified) {
+                userMail = payload.email
+            }
+
+        } else {
+            throw Exception("Token cannot be verified")
+        }
+        if (userMail.isNullOrBlank()) {
+            throw Exception("Token cannot be verified")
+        }
```
Comment (1MPULSEONE) [inline effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/TokenVerifier.kt:37]: In order to get rid of unnecessary nesting, the token check can be put above it

File: effectiveOfficeBackend/src/main/kotlin/office/effective/model/IntegrationModel.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+data class IntegrationModel(var name: String, var valueStr: String) {
+    var id: UUID? = null
+    var iconUrl: String = ""
+
+    constructor(
+        _id: UUID,
+        _name: String,
+        _valueStr: String,
+        _iconUrl: String
```
Comment (1MPULSEONE) [inline effectiveOfficeBackend/src/main/kotlin/office/effective/model/IntegrationModel.kt:13]: You don`t need to use underlines here, you use "this" anyway.

Comment (1MPULSEONE) [reply effectiveOfficeBackend/src/main/kotlin/office/effective/model/IntegrationModel.kt:13]: Read about constructors in kotlin, you can write this in one line, just using primary constructor, not secondary

File: effectiveOfficeBackend/src/main/kotlin/office/effective/model/UserModel.kt [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        _avatarURL: String?,
+        _integrations: Set<IntegrationModel>?
+    ) : this(_fullName) {
+        this.id = _id;
+        this.tag = _tag;
+        this.active = _active
+        this.role = _role
+        this.avatarURL = _avatarURL
+        this.integrations = _integrations
+    }
```
Comment (1MPULSEONE) [inline effectiveOfficeBackend/src/main/kotlin/office/effective/model/UserModel.kt:30]: Read about constructors in kotlin, you can write this in one line, just using primary constructor, not secondary

File: effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/converters/IntegrationModelEntityConverter.kt [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+import office.effective.feature.auth.repository.IntegrationEntity
+import office.effective.model.IntegrationModel
+
+class IntegrationModelEntityConverter {
+    fun modelToEntity(integrationModel: IntegrationModel): IntegrationEntity {
+        return IntegrationEntity {
+            id = integrationModel.id
+            name = integrationModel.name;
+            iconUrl = integrationModel.iconUrl
```
Comment (1MPULSEONE) [inline effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/converters/IntegrationModelEntityConverter.kt:11]: I think named arguments are redundant here

File: effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/converters/IntegrationModelEntityConverter.kt [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        }
+    }
+
+    fun entityToModel(integrationEntity: IntegrationEntity, valueStr: String): IntegrationModel {
+        return IntegrationModel(
+            _id = integrationEntity.id!!,
+            _iconUrl = integrationEntity.iconUrl,
+            _name = integrationEntity.name,
+            _valueStr = valueStr
```
Comment (1MPULSEONE) [inline effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/converters/IntegrationModelEntityConverter.kt:20]: Poor work with constructors

File: effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/converters/UserModelEntityConverter.kt [hunk 6]
```diff
@@ -0,0 +1,27 @@
+package office.effective.feature.auth.converters
+
+import office.effective.feature.auth.repository.UserEntity
+import office.effective.model.UserModel
+
+class UserModelEntityConverter {
+    fun modelToEntity(userModel: UserModel): UserEntity {
+        var res = UserEntity()
+        res.id = userModel.id!!
```
Comment (1MPULSEONE) [inline effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/converters/UserModelEntityConverter.kt:9]: Try to get rid of "!!". You should use such an operator only if you are 100% sure that you won't have null here, but the compiler doesn't have enough information to realize that. I'm not sure if this is your case

File: effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/repository/UserRepository.kt [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+import org.ktorm.entity.*
+import java.util.*
+
+class UserRepository {
+    private val db: Database = GlobalContext.get().get()
+    private val converter: UserModelEntityConverter = GlobalContext.get().get()
+
+
+    fun existsById(userId: UUID): Boolean {
+        return db.users.count { it.id eq userId } > 0
```
Comment (1MPULSEONE) [inline effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/repository/UserRepository.kt:22]: Better to use find

File: effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/repository/UserRepository.kt [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    fun findById(userId: UUID): UserModel {
+        val userEnt: UserEntity =
+            db.users.find { it.id eq userId } ?: throw UserNotFoundException("DB sync error")
+        val integrations = findSetOfIntegrationsByUser(userEnt.id!!)
```
Comment (1MPULSEONE) [inline effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/repository/UserRepository.kt:28]: !! is dangerous

File: effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/repository/UserRepository.kt [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+    fun findByEmail(email: String): UserModel {
+        val integrationUserEntity: UserIntegrationEntity = db.usersinegrations.find { it.valueStr eq email }
+            ?: throw UserIntegrationNotFoundException("not such email");
+        return findById(integrationUserEntity.userId.id)
+    }
+
+    fun findIntegrationById(id: UUID): IntegrationEntity {
+        return db.integrations.find { it.id eq id }
+            ?: throw IntegrationNotFoundException("There are no integration with such id")
```
Comment (1MPULSEONE) [inline effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/repository/UserRepository.kt:60]: Bring the description of the errors to one style (capitalized, etc.)

File: effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/repository/UserRepository.kt [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    fun findSetOfIntegrationsByUser(userId: UUID): MutableSet<IntegrationModel> {
+
+        val usersIntegrationsSet: Set<UserIntegrationEntity> =
+            db.usersinegrations.filter { it.userId eq userId }.toSet()
+        var modelsSet: MutableSet<IntegrationModel> = mutableSetOf()
+
+        usersIntegrationsSet.forEach {
+            val integrationEntity = findIntegrationById(it.integrationId.id!!)
```
Comment (1MPULSEONE) [inline effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/repository/UserRepository.kt:70]: !! is dangerous

File: effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/repository/UserRepository.kt [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+            modelsSet.add(
+                IntegrationModel(
+                    _id = integrationEntity.id!!,
+                    _name = integrationEntity.name,
+                    _valueStr = it.valueStr,
+                    _iconUrl = integrationEntity.iconUrl
```
Comment (1MPULSEONE) [inline effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/repository/UserRepository.kt:76]: poor work with constructor

File: effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/routing/AuthRouting.kt [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+fun Route.authRoutingFun() {
+    val service: IUserService = GlobalContext.get().get()
+    authenticate("auth-oauth-google") {
+        get("/login") {
+
+        }
+        get("/callback") {
+            try {
+
```
Comment (1MPULSEONE) [inline effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/routing/AuthRouting.kt:24]: Lint your code in IDE (ctrl+alt+l)

File: effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/routing/AuthRouting.kt [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+            call.respondRedirect("http://localhost:8080/callback")
+        }
+    }
+    get("/user/login") {
+        val tokenStr = call.request.header("id_token") ?: call.response.status(HttpStatusCode.Forbidden)
+        val userDTO: UserDTO = service.getUserByToken(tokenStr as String)
+//
+    }
+
+}
```
Comment (1MPULSEONE) [inline effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/routing/AuthRouting.kt:64]: Dirty code, with comments and prints like this should never make it into pullrequest.

File: effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/service/UserService.kt [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+        TODO("Not yet implemented")
+    }
+
+    override fun getUserById(userIdStr: String): UserDTO {
+        TODO("Not yet implemented")
+    }
+
+    override fun alterUser(user: UserDTO): UserDTO {
+        TODO("Not yet implemented")
+    }
```
Comment (1MPULSEONE) [inline effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/service/UserService.kt:26]: Write next to the *TODO* who they apply to (and a link to the board with the task if possible).

**Review by 1MPULSEONE — CHANGES_REQUESTED** [review]

**Comment by 1MPULSEONE** [comment]
It's worth writing a description for any pull request, read about git-flow in the repository documentation

File: effectiveOfficeBackend/src/main/kotlin/office/effective/Application.kt [hunk 15]
```diff
@@ -17,9 +27,27 @@ fun main() {
...
+        oauth("auth-oauth-google") {
+            urlProvider = { "http://localhost:8080/callback" }
+            providerLookup = {
+                OAuthServerSettings.OAuth2ServerSettings(
+                    name = "google",
+                    authorizeUrl = "https://accounts.google.com/o/oauth2/auth",
+                    accessTokenUrl = "https://accounts.google.com/o/oauth2/token",
+                    requestMethod = HttpMethod.Get,
+                    clientId = "27867488691-gc95ln5jad3i84dcmu3dd9ls1s4hvm9c.apps.googleusercontent.com", /*System.getenv("GOOGLE_CLIENT_ID")*/
+                    clientSecret = "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",//System.getenv("GOOGLE_CLIENT_SECRET"),
```
[the remark this answers is not in the document]
Comment (kiselev-danil) [reply effectiveOfficeBackend/src/main/kotlin/office/effective/Application.kt]: Fixed

File: effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/routing/AuthRouting.kt (hunk 13 above)
Comment (kiselev-danil) [reply effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/routing/AuthRouting.kt]: Deleted comments

File: effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/routing/AuthRouting.kt (hunk 12 above)
Comment (kiselev-danil) [reply effectiveOfficeBackend/src/main/kotlin/office/effective/feature/auth/routing/AuthRouting.kt:25]: Done

**Pull request closed without merging**